File: Assets/src/model/grasses/VibrantIvy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using UnityEngine;

[Serializable]
[ObjectInfo("vibrant-ivy")]
public class VibrantIvy : Grass, IActorEnterHandler, IAttackHandler, IActorLeaveHandler {
  public override string description =>
    $"Camouflages the player.\nCreatures that would chase you will move randomly instead, and you will not wake sleeping creatures.\nMoving or attacking from within the Vibrant Ivy destroys it.";

  public static bool CanOccupy(Tile tile) {
    var floor = tile.floor;
    // hugging at least one 4-neighbor wall
    var cardinalNeighbors = floor.GetCardinalNeighbors(tile.pos);
    var isHuggingWall = cardinalNeighbors.Any((pos) => pos is Wall);
    var isGround = tile is Ground || tile is Water;
    var isNotIvy = !(tile.grass is VibrantIvy);
    
    return isHuggingWall && isGround && isNotIvy;
  }

  [OptionalField] // added 1.11.0
  private int stacks;
  public int Stacks => stacks;

  public VibrantIvy(Vector2Int pos) : base(pos) {
    BodyModifier = this;
  }

  protected override void HandleEnterFloor() {
    ComputeStacks();
    if (actor is Player player) {
      player.statuses.Add(new CamouflagedStatus());
    }
  }

  protected override void HandleLeaveFloor() {
    if (actor is Player player) {
      player.statuses.RemoveOfType<CamouflagedStatus>();
    }
  }

  public void HandleActorEnter(Actor who) {
    if (who is Player player) {
      player.statuses.Add(new CamouflagedStatus());
      OnNoteworthyAction();
    }
  }

  public void OnAttack(int damage, Body target) {
    if (actor is Player player) {
      LoseStack(player);
    }
  }

  public void HandleActorLeave(Actor who) {
    if (who is Player player) {
      LoseStack(player);
    }
  }

  private void LoseStack(Player player) {
    stacks--;
    if (stacks == 0) {
      Kill(player);
    }
  }

  public void ComputeStacks() {
    stacks = floor.GetCardinalNeighbors(pos).Where(t => t is Wall).Count();
  }
}

[Serializable]
[ObjectInfo("vibrant-ivy")]
internal class CamouflagedStatus : Status, IPlayerCamouflage {
  public CamouflagedStatus() {
  }

  public override void Start() {
    actor.floor.RecomputeVisibility();
  }

  public override void End() {
    actor.floor.RecomputeVisibility();
  }

  public override bool Consume(Status other) => true;

  public override string Info() =>
    "Creatures that would chase you will move randomly instead, and you will not wake sleeping creatures.\nMoving or attacking from within the Vibrant Ivy destroys it.";

  public override void Step() {
    if (!(actor?.grass is VibrantIvy)) {
      Remove();
    }
  }
}

public interface IPlayerCamouflage {}

File: web/src/model/grasses/VibrantIvy.ts
import { Grass } from './Grass';
import { Status } from '../Status';
import {
  ACTOR_ENTER_HANDLER,
  ACTOR_LEAVE_HANDLER,
  type IActorEnterHandler,
  type IActorLeaveHandler,
} from '../../core/types';
import { ATTACK_HANDLER, type IAttackHandler } from '../Actor';
import { Ground, Wall, Water } from '../Tile';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';
import type { Body } from '../Body';

const CAMOUFLAGE = Symbol.for('IPlayerCamouflage');

/**
 * Camouflages the player. Stacks = adjacent cardinal walls.
 * Moving or attacking from within reduces stacks; at 0 the grass dies.
 * Port of C# VibrantIvy.cs.
 */
export class VibrantIvy extends Grass implements IActorEnterHandler, IAttackHandler, IActorLeaveHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;
  readonly [ACTOR_LEAVE_HANDLER] = true as const;
  readonly [ATTACK_HANDLER] = true as const;

  private _stacks = 0;

  get stacks(): number {
    return this._stacks;
  }

  get bodyModifier(): object | null {
    return this;
  }

  constructor(pos: Vector2Int) {
    super(pos);
  }

  static canOccupy(tile: Tile): boolean {
    const floor = tile.floor!;
    const isHuggingWall = floor.getCardinalNeighbors(tile.pos).some(t => t instanceof Wall);
    const isGround = tile instanceof Ground || tile instanceof Water;
    const isNotIvy = !(tile.grass instanceof VibrantIvy);
    return isHuggingWall && isGround && isNotIvy;
  }

  protected handleEnterFloor(): void {
    this.computeStacks();
    const standing = this.floor?.bodies.get(this.pos);
    if (standing?.constructor.name === 'Player') {
      (standing as any).statuses.add(new CamouflagedStatus());
    }
  }

  protected handleLeaveFloor(): void {
    const standing = this.floor?.bodies.get(this.pos);
    if (standing?.constructor.name === 'Player') {
      (standing as any).statuses.removeOfType(CamouflagedStatus);
    }
  }

  handleActorEnter(who: any): void {
    if (who.constructor.name === 'Player') {
      who.statuses.add(new CamouflagedStatus());
      this.onNoteworthyAction();
    }
  }

  onAttack(_damage: number, _target: Body): void {
    const standing = this.floor?.bodies.get(this.pos);
    if (standing?.constructor.name === 'Player') {
      this.loseStack(standing as any);
    }
  }

  handleActorLeave(who: any): void {
    if (who.constructor.name === 'Player') {
      this.loseStack(who);
    }
  }

  private loseStack(player: any): void {
    this._stacks--;
    if (this._stacks === 0) {
      this.kill(player);
    }
  }

  computeStacks(): void {
    this._stacks = this.floor!.getCardinalNeighbors(this.pos).filter(t => t instanceof Wall).length;
  }
}

/**
 * Player is camouflaged — enemies won't chase, sleeping creatures won't wake.
 * Removes itself when the player is no longer standing on VibrantIvy.
 * Port of C# CamouflagedStatus from VibrantIvy.cs.
 */
export class CamouflagedStatus extends Status {
  readonly [CAMOUFLAGE] = true as const;

  Start(): void {
    this.actor?.floor?.recomputeVisibility();
  }

  End(): void {
    this.actor?.floor?.recomputeVisibility();
  }

  Consume(_other: Status): boolean {
    return true;
  }

  Step(): void {
    if (!(this.actor?.grass instanceof VibrantIvy)) {
      this.Remove();
    }
  }
}

entityRegistry.register('VibrantIvy', VibrantIvy);
